zerver/tests/test_message_fetch.py

```diff
@@ -593,31 +592,243 @@ def _build_query(self, term: Dict[str, Any]) -> Select:
+        self.assertFalse(
+            narrow_filter(
+                message={"display_recipient": "social", "type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["sender", "hamlet@zulip.com"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"sender_email": "hamlet@zulip.com"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"sender_email": "cordelia@zulip.com"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "dm"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"type": "private"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "private"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"type": "private"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "starred"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["alerted"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "alerted"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["mentioned"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "mentioned"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["mentioned"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "unread"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["read"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "resolved"]])
```

Comment:
It's possible it might be nicer to use a different name for each of these, like `is_resolved_filter`; I'll leave that for you to submit a follow-up PR if you like the idea.

Reply:
I like the idea, but I'll probably re-visit this code when I get to supporting negated queries, so I'll keep the idea in the back of my mind for now.